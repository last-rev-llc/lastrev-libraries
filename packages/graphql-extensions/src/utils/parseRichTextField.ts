import { documentToPlainTextString } from '@contentful/rich-text-plain-text-renderer';
import { Document } from '@contentful/rich-text-types';
import { getLocalizedField } from '@last-rev/graphql-contentful-core';
import { ApolloContext } from '@last-rev/types';
import { Entry } from 'contentful';

const collectLinks = (node: any) => {
  const entriesLinks: any[] = [];
  const {
    data: { target },
    content
  } = node;
  // DFS collect children links
  if (content?.length) {
    content.forEach(collectLinks);
  }
  if (target?.sys?.type === 'Link' && target?.sys?.linkType === 'Entry') {
    entriesLinks.push(target);
  }
  return entriesLinks;
};

async function parseEntry(entry: any, ctx: ApolloContext) {
  if (!entry || !entry?.sys?.id) return '';
  switch (entry.sys.contentType.sys.id) {
    case 'section': {
      const contentIds = (getLocalizedField(entry.fields, 'contents', ctx) || []).map((c: Entry<any>) => c.sys.id);
      const loaded = await ctx.loaders.entryLoader.loadMany(
        contentIds.map((id: string) => ({ id, preview: !!ctx.preview }))
      );
      const parsed: string[] = await Promise.all(
        loaded.filter((l) => !!(l as Entry<any>)?.sys?.id).map((e) => parseEntry(e as Entry<any>, ctx))
      );
      return parsed.join(' ');
    }
    case 'text': {
      const body = getLocalizedField(entry.fields, 'body', ctx);
      return await parseRichTextField(body, ctx);
    }
    case 'table': {
      return getLocalizedField(entry.fields, 'table', ctx) || '';
    }
    default:
      return '';
  }
}

async function parseRichTextField(richText: Document | any | undefined, ctx: ApolloContext): Promise<string> {
  if (!richText) return '';

  const entriesLinks = collectLinks(richText);

  const entries = await ctx.loaders.entryLoader.loadMany(entriesLinks.map((id) => ({ id, preview: !!ctx.preview })));

  const entryVals = await Promise.all(entries.map(async (r) => parseEntry(r, ctx)));

  return [documentToPlainTextString(richText), ...entryVals].join(' ');
}

export default parseRichTextField;
