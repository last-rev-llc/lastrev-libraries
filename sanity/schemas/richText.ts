export default {
  name: 'richText',
  title: 'Rich Text',
  type: 'array',
  of: [{ type: 'block' }]
}
