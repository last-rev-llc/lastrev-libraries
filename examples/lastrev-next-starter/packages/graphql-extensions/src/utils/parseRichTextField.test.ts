import parseRichTextField from './parseRichTextField';
import richText from './richText.mock';
import { BLOCKS } from '@contentful/rich-text-types';

describe('parseRichTextField.ts', () => {
  it('returns an empty array when there is no content', () => {
    const parsed = parseRichTextField(undefined, { sectionDelimiter: BLOCKS.HEADING_2 });
    expect(parsed).toEqual([]);
  });
  it('works when there is content', () => {
    const parsed = parseRichTextField(richText, { sectionDelimiter: BLOCKS.HEADING_2 });
    expect(parsed).toMatchSnapshot();
  });
});
