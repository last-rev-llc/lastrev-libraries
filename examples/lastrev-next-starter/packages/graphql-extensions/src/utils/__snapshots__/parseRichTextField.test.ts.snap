// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseRichTextField.ts works when there is content 1`] = `
Array [
  Object {
    "content": "Enhanced Some Text does something else.",
    "section": "",
  },
  Object {
    "content": "Lorem ipsumTest LinkThis is more text here. ",
    "section": "A Section Header",
  },
]
`;
